File: 2130-MaximumTwinSumofaLinkedList/2130-MaximumTwinSumofaLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        dummy = ListNode(val=0)
        dummy.next = head
        prev = dummy

        slow = head
        fast = head
        curr = head

        ans = 0

        while curr:
            if fast:
                #set head and tail of half of the list
                sub_head = curr
                #reverse first half of the list
                while fast:
                    fast = fast.next.next
                    next = curr.next
                    curr.next = prev
                    prev = curr
                    curr = next
                #connect the reverse half with the second half
                sub_head.next = curr
                dummy.next = prev

            # calculate sum and update the max
            ans = max(ans, prev.val + curr.val)
            prev = prev.next
            curr = curr.next
        
        return ans

        

File: 2130-MaximumTwinSumofaLinkedList/2130-MaximumTwinSumofaLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function(head) {
    let fast = head;
    
    let curr = head;
    let dummy = new ListNode();
    dummy.next = head;
    let prev = dummy;
    let ans = 0


    while(curr){
        if(fast){
            let subHead = curr;
            while(fast){
                fast = fast.next.next;
                let next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }
            subHead.next  = curr;
            dummy.next = prev;
        }
        ans = Math.max(ans, prev.val + curr.val);
        prev = prev.next;
        curr = curr.next;
    }
    return ans;
    
};